mapping_parenting_tech/analysis/prototyping/play_store/get_play_store_apps.py

```diff
@@ -7,141 +7,413 @@
+            )
+
+            # if we just grabbed a full 200 reviews, there are (probably) more to get
+            more_to_get = True if len(review_fetch) == 200 else False
+
+            # process what we've just downloaded:
+            # 1. add the add id to each review
+            # 2. remove the userName and userImage
+            # 3. add to the target file if we don't already have it
+            with open(target_file, "at", newline="") as csv_file:
+                csv_writer = csv.DictWriter(csv_file, fieldnames=field_names)
+                for review in review_fetch:
+                    review.update({"appId": app_id})
+                    del review["userName"]
+                    del review["userImage"]
+                    if review["reviewId"] not in existing_review_ids:
+                        csv_writer.writerow(review)
+                        existing_review_ids.append(review["reviewId"])
+
+            # save a snapshot of where we are - the app_id and continuation_token - so we can resume later
+            with open(log_file, "rb+") as log_file_handle:
+                if first_pass:
+                    log_dump = {
+                        app_id: {
+                            "completed": False,
+                            "continuation_token": continuation_token,
+                            "latest_review_id": review_fetch[0]["reviewId"],
+                            "latest_review_time": review_fetch[0]["at"],
+                            "downloaded": len(review_fetch),
+                        }
+                    }
+                else:
+                    log_dump[app_id].update(
+                        {
+                            "continuation_token": continuation_token,
+                            "downloaded": (
+                                log_dump[app_id]["downloaded"] + len(review_fetch)
+                            ),
+                        }
+                    )
+                log_info.update(log_dump)
+                pickle.dump(log_info, log_file_handle)
+
+            app_review_count += len(review_fetch)
+
+            if not run_quietly:
+                print(
+                    f"Retrieved {app_review_count} new reviews for {app_id}; fetched {log_dump[app_id]['downloaded']} in total)          ",
+                    end="\r",
+                )
+
+            first_pass = False
+            if force_download:
+                keep_going = True
+
+        # if we've reached here and keep_going is true it's because we've downloaded all the results, so update logfile accordingly
+        if keep_going:
+            with open(log_file, "rb+") as log_file_handle:
+                log_dump[app_id].update({"completed": True, "continuation_token": None})
+                log_info.update(log_dump)
+                pickle.dump(log_info, log_file_handle)
+        if not run_quietly:
+            print(f"\n{app_id}: done")
+
+    return fetched_reviews
```

Comment:
This is quite a complex function - perhaps it's possible to aim towards the so-called [single responsibility principle](https://en.wikipedia.org/wiki/Single-responsibility_principle) and try factoring out some of the csv and log file reading and writing operations as stand-alone functions? Once it's factored out, this particular function would be more compact and it would also be easier to see what it's doing, ie I can imagine something like this
```
def get_playstore_app_list_reviews(...):
    ...
    check_target_file_exists(filename)
    check_log_file_exists(filename)
    for app_id in app_id_list:
        ...
        check_app_in_log_file(app_id)
        while ...
            get_playstore_app_reviews()
            process_app_data()
            save_app_data()
            record_log()
            ... 
```
This benefits are even more apparent when you start writing unit tests.

For now, however, it's ok to perhaps keep the refactoring task as another github issue for later.

Reply:
Have tried to do this as best I can - there are necessarily still steps between these functions to update local variables, make decisions, etc.